### PR TITLE
New AddressGen for xdma

### PR DESCRIPTION
**Pull request by IveanEx**
As mentioned in Issue #225 , the Address Generator is suspected to be the main cause of the long critical path. This PR redesign the AGU of XDMA while keeping the interface compatible to the original AGU. 

The key changes include: 
- Remove the original design where there is only one centralized counter, and each dimension is calculated based on division of centralized counter's value. This removes the division in the original addressgen. 
- Early addition of stride: introduce a programmable counter that the step size equals to stride. This removes the multiplication in the original addressgen
- Log2 addition in address offset: The addition of address offset is replaced by an automatic adder tree to reduce the critical path. 
- Bitwidth control: instead of letting Chisel to automatically infer the bitwidth of operand, the bitwidth is explicitly defined: the bitwidth is log2(tcdmSize). And this is done by automatic reading the cfg. 
- Late addition of base address: the base address is added at the very end, so most of the adder has width of just 17 bit (128KB condition).

### CODE REVIEW

**Review by rgantonio — CHANGES_REQUESTED** [review]
OK!

I just wonder but, so the change to have a new address generation that is not multiplication dependent was to add new features? What happened to the one that has multiplication or division?

Also, please resolve comments and questions before proceeding! I will approve after you ping me again 🤣 

Thanks @IveanEx !

**Comment by IveanEx** [comment]
> OK!
> 
> I just wonder but, so the change to have a new address generation that is not multiplication dependent was to add new features? What happened to the one that has multiplication or division?
> 
> Also, please resolve comments and questions before proceeding! I will approve after you ping me again 🤣
> 
> Thanks @IveanEx !

When designing the original xdma, the functional correction is the only target. Thus, some function is realized, although quite descent from the viewers' side, it is not friendly to the hardware generation. The new addressgen has no new functions, but totally rework on how the new address offset is calculated😂

File: hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala:80]: I double-checked the implementation here. There are two reset pins, one is at *top.reset*, where is an asynchronous reset to bring the circuit back to normal state after power up; and another one is *top.io.reset*, where is a synchronous reset for the correct functioning.

File: hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala:90]: Chisel is based on Scala. So, when writing Chisel, you are actually writing two languages simultaneously! I think this view is very important for people writing Chisel. 

For your question, **if / else** is *scala* condition, they are evaluated during the generation phase. **when / otherwise** is *chisel* condition, and they become the actual hardware implementation later. 

Other examples are **switch...case...** in scala and **case...is...** in chisel, **!=** in scala and **=/=** in chisel etc.

File: hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala:274]: Although the symbol of Scala multiplier and Chisel Multiplier are the same, Scala multiplier's operand belongs to **Integer** class while Chisel multiplier's operand belongs to **Numeric** trait. By setting the operand correctly, Scala can do operator overloading that fully represents my design choice.

File: hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala]: Improved

File: hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala:303]: Improved

File: hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala:235]: The original AddressGenUnitParam still stays there. The only difference is that the new parameter also has the size of the memory. This information is come at elsewhere, if I totally replace it, then there are two places to define the size, which can easily become the source of error.

File: hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala
Comment (rgantonio) [reply hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala:90]: Aha! Cool okay, I understand! Thanks 😄

**Review by rgantonio — APPROVED** [review]
OK!

File: hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala:265]: Later lines (ten or twenty lines? ) haha only when after all signals are declared, then I can do the connection.

File: hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala:80]: The reset at the top, is managed by chisel implicitly. At Line 62, ***val value = RegNext(nextValue, 0.U)*** means that this register should be reset to 0.U when reset signal (top one) is pulled up. On the other hand, the synchronous one, is used explicitly (which is the one you see in the code).

**Pull request merged**
